src/nwb_conversion_tools/tools/nwb_helpers.py

```diff
@@ -117,52 +117,61 @@ def add_device_from_metadata(nwbfile: NWBFile, modality: str = "Ecephys", metada
-        Path for where to write the NWBFile.
-    metadata: dict, optional
-        Metadata dictionary with information used to create the NWBFile when one does not exist or overwrite=True.
+        Path for where to write or load (if overwrite=False) the NWBFile.
+        If specified, this context will always write to the this location.
     nwbfile: NWBFile, optional
         An in-memory NWBFile object to write to the location.
+    metadata: dict, optional
+        Metadata dictionary with information used to create the NWBFile when one does not exist or overwrite=True.
     overwrite: bool, optional
         Whether nor not to overwrite the NWBFile if one exists at the nwbfile_path.
         The default is False (append mode).
+    verbose: bool, optional
+        If 'nwbfile_path' is specified, informs user after a successful write operation.
+        The default is True.
     """
-    assert not (overwrite is False and Path(nwbfile_path).exists() and nwbfile is not None), (
+    nwbfile_path_in = Path(nwbfile_path) if nwbfile_path else None
+    assert not (nwbfile_path is None and nwbfile is None and metadata is None), (
+        "You must specify either an 'nwbfile_path', or an in-memory 'nwbfile' object, "
+        "or provide the metadata for creating one."
+    )
+    assert not (overwrite is False and nwbfile_path_in and nwbfile_path_in.exists() and nwbfile is not None), (
         "'nwbfile_path' exists at location, 'overwrite' is False (append mode), but an in-memory 'nwbfile' object was "
         "passed! Cannot reconcile which nwbfile object to write."
     )
-    if metadata is not None and nwbfile is not None:
-        warn(
-            "Passing an in-memory NWBFile object, but also passing metadata for building a fresh NWBFile. "
-            "Metadata will be ignored."
-        )
-    if metadata is not None and overwrite is False and Path(nwbfile_path).exists():
-        warn(
-            f"Writing to 'nwbfile_path' ({nwbfile_path}) in append mode, but also passing metadata for building a "
-            "fresh NWBFile. Metadata will be ignored and the existing file will be appended."
-        )
 
-    load_kwargs = dict(path=nwbfile_path)
-    if Path(nwbfile_path).is_file() and not overwrite:
-        load_kwargs.update(mode="r+", load_namespaces=True)
-    else:
-        load_kwargs.update(mode="w")
-    io = NWBHDF5IO(**load_kwargs)
+    load_kwargs = dict()
+    if nwbfile_path:
+        load_kwargs.update(path=nwbfile_path)
+        if nwbfile_path_in.is_file() and not overwrite:
+            load_kwargs.update(mode="r+", load_namespaces=True)
+        else:
+            load_kwargs.update(mode="w")
+        io = NWBHDF5IO(**load_kwargs)
     try:
-        if load_kwargs["mode"] == "r+":
+        if load_kwargs.get("mode", "") == "r+":
             nwbfile = io.read()
         elif nwbfile is None:
             nwbfile = make_nwbfile_from_metadata(metadata=metadata)
         yield nwbfile
     finally:
```

Comment:
Not really sure what kind of comments you'd want for this that would not just be the name of the function itself. 164+165 tries to load an nwbfile, if that is the intended usage mode, other case of 166+167 is to make otherwise make a fresh one

Reply:
I quoted the wrong line. I was talking about what happens after `169`.

Reply:
Still not really sure what comments you'd want - a `finally` clause is always called after success or capture, if a path is specified it tried to write the file and always closes the `io` no matter how that goes

Reply:
Right, so that last comment is that I was missing as I don't think I understand it fully until now. Could it be "if the `nwbfile_path` exists always write and close the `io` object when exiting the context manager. 

The threefold structure if I get it right is:
1) if `nwbfile_path` exists defines an `io` object for writing to the file. (157)
2) Load the nwbfile to memory either from the `nwbfile_path` or build it from metadata. (164)
3)  Could it be "if the `nwbfile_path` exists always write and close the `io` object when exiting the context manager. (171).

It might be that someone more experience with context managers can parse this directly from the code making the comments redundant. I flagged it because I did not understand it even after we have discussed this in person before.